### DISCUSSION TITLE
[general] Can't interactively render tar_quarto notebook in a subdirectory

### DISCUSSION THREAD

**Discussion by shamindras** [ropensci/targets#1170] [Help] [answered]
### Help

- [X] I understand and agree to <https://books.ropensci.org/targets/help.html>.

### Description

Hi,

I'm a new {targets} user, and really enjoying it! I'm running into a problem with interactively knitting a quarto notebook that is in a different directory than the `_targets.R` file.

To be precise, my reprex directory looks like this minimal structure:

<details><summary>Directory structure</summary>

```bash
.
├── _targets
│   ├── meta
│   ├── objects
│   └── user
├── _targets.R
├── documents
│   ├── test-targets-quarto_b.html
│   ├── test-targets-quarto_b.qmd
│   └── test-targets-quarto_b_files
├── renv
│   ├── activate.R
│   ├── library
│   ├── settings.json
│   └── staging
├── renv.lock
├── run.R
├── run.sh
├── setup-repo.R
├── targets_quarto_test.Rproj
├── test-targets-quarto_a.html
├── test-targets-quarto_a.qmd
└── test-targets-quarto_a_files
    └── libs
```

</details>

The `_targets.R` file contains the following basic user manual code with a couple of `tar_quarto()` calls:

<details><summary>_targets.R</summary>

```R
# Created by use_targets().
# Follow the comments below to fill in this target script.
# Then follow the manual to check and run the pipeline:
#   https://books.ropensci.org/targets/walkthrough.html#inspect-the-pipeline

# Load packages required to define the pipeline:
library(targets)
library(tarchetypes) # Load other packages as needed.

# Set target options:
tar_option_set(
  packages = c("tibble", "here", "quarto") # packages that your targets need to run
  # format = "qs", # Optionally set the default storage format. qs is fast.
  #
  # For distributed computing in tar_make(), supply a {crew} controller
  # as discussed at https://books.ropensci.org/targets/crew.html.
  # Choose a controller that suits your needs. For example, the following
  # sets a controller with 2 workers which will run as local R processes:
  #
  #   controller = crew::crew_controller_local(workers = 2)
  #
  # Alternatively, if you want workers to run on a high-performance computing
  # cluster, select a controller from the {crew.cluster} package. The following
  # example is a controller for Sun Grid Engine (SGE).
  # 
  #   controller = crew.cluster::crew_controller_sge(
  #     workers = 50,
  #     # Many clusters install R as an environment module, and you can load it
  #     # with the script_lines argument. To select a specific verison of R,
  #     # you may need to include a version string, e.g. "module load R/4.3.0".
  #     # Check with your system administrator if you are unsure.
  #     script_lines = "module load R"
  #   )
  #
  # Set other options as needed.
)

# tar_make_clustermq() is an older (pre-{crew}) way to do distributed computing
# in {targets}, and its configuration for your machine is below.
options(clustermq.scheduler = "multicore")

# tar_make_future() is an older (pre-{crew}) way to do distributed computing
# in {targets}, and its configuration for your machine is below.
# Install packages {{future}}, {{future.callr}}, and {{future.batchtools}} to allow use_targets() to configure tar_make_future() options.

# Run the R scripts in the R/ folder with your custom functions:
# tar_source()
# source("other_functions.R") # Source other scripts as needed.

# Replace the target list below with your own:
list(
  tar_target(
    name = data,
    command = tibble(x = rnorm(100), y = rnorm(100))
    # format = "feather" # efficient storage for large data frames
  ),
  tar_target(
    name = model,
    command = coefficients(lm(y ~ x, data = data))
  ),
  # This report renders as expected
  tar_quarto(name = reporta, path = here::here("test-targets-quarto_a.qmd")),
  # This report does not render as expected
  tar_quarto(name = reportb, path = here::here("documents/test-targets-quarto_b.qmd"))
)
```

</details>

The key thing to note is that there are 2 quarto docs, with **identical content**, but in **two separate locations**.
They are `./test-targets-quarto_a.qmd` and `documents/test-targets-quarto_b.qmd`. The latter is importantly in the `documents` subdirectory, which is not at the same level as `_targets.R`. The contents of these files is:

<details><summary>Contents of quarto files</summary>

<pre>
---
title: "This doc is in the same dir as _targets.R"
format: html
editor: source
---

## Quarto

Quarto enables you to weave together content and executable code into a finished document. To learn more about Quarto <https://quarto.org>.

```{r}
library(targets)
tar_load(data)
data
```
</pre>

</details>

Importantly, running `tar_make()` works here and produces the following output:

```
> targets::tar_make()
- The project is out-of-sync -- use `renv::status()` for details.
✔ skip target data
▶ start target reporta
● built target reporta [3.155 seconds]
✔ skip target reportb
✔ skip target model
▶ end pipeline [3.272 seconds]
```

And when I interactively click "render" inside `test-targets-quarto_a.qmd`, it renders in the RStudio IDE viewer pane. So `test-targets-quarto_a.qmd` really gives me the best of both worlds. Using `tar_make()` to render via console as needed when I save it with updates, and the ability to interactively render in the RStudie IDE viewer pane. 

However, when I try to _interactively render_ `documents/test-targets-quarto_b.qmd`, it results in the following error:

```


processing file: test-targets-quarto_b.qmd
  |...................................                 |  67% [unnamed-chunk-1]
Quitting from lines 12-15 [unnamed-chunk-1] (test-targets-quarto_b.qmd)
Error:
! targets data store _targets not found. Utility functions like tar_read() and tar_load() require a pre-existing targets data store (default: _targets/) created by tar_make(), tar_make_clustermq(), or tar_make_future(). Details: https://books.ropensci.org/targets/data.html
                                                                                                            
Execution halted

```

The issue seems to be that it is interactively looking for the `_targets.R` file, but it is not in the same directory. I'm not sure how to fix this, since iteratively and interactively writing and rendering the quarto notebooks is a key part of my workflow.

Is there a way to fix this without moving `test-targets-quarto_b.qmd` to the root folder? I'd like to keep the existing directory structure and enjoy the ability to **both** render `documents/test-targets-quarto_b.qmd` interactively in the RStudio viewer, and also to be able render it using `tar_make()` as needed.

**Comment by shamindras** [comment]
I meant to clarify, that the above reprex is essentially adapted from the [user manual](https://books.ropensci.org/targets/literate-programming.html) guide on Literate programming, just copying over one of the quarto notebooks to another subdirectory, and giving it a new target name.

**Comment by wlandau** [comment, marked as answer]
Your example works on my end, so I think you may need to upgrade your version of `tarchetypes`.

**Reply by shamindras** [reply]
Wow - this worked! Thanks @wlandau .

Surprisingly, I was just about to post this alternative fix I discovered by reading more deeply into the [target configuration settings](https://docs.ropensci.org/targets/reference/tar_config_set.html):

```{r}
library(targets)
tar_config_set(store = here::here("_/targets/")) # new line added to point to correct `_targets` store at root project dir using {here}
tar_load(data)
data
```

Would it still be a more robust `{targets}` style to _always_ include this one-line `tar_config_set(store = here::here("_/targets/"))` in each file that depends on targets? Especially if the file is located in another directory than `_targets.R`? Or does the latest version of `tarchetypes` essentially do this same thing for each file which contains a `tar_` function call?

**Reply by wlandau** [reply]
I am surprised you needed to set the store in the most current version of `tarchetypes`, but that will definitely work, and it's not a bad practice.

**Reply by shamindras** [reply]
@wlandau I should have clarified better. The latest version of `tarchetypes` definitely fixed this issue. I just upgraded and re-ran and it all worked (after removing `tar_config_set`).

My question on adding the `tar_config_set(store = here::here("_/targets/"))` as a general rule was just to check whether it would make `targets` pipelines even more robust to these issues (even after upgrading to the latest `tarchetypes` version). Per your recommendation I may continue to add this line for safety.
